### PR TITLE
Proper client timeout detection

### PR DESCRIPTION
**Pull request by Treeston**
Properly detect client timeout when logged into a character after a configurable time (default 10s) has passed without the client sending any packets.

Fixes issues with crashed clients leaving characters in the world for a very long time (default 15 minutes), as well as edge case exploits involving intentionally pausing client execution for some amount of time.

### CODE REVIEW

**Comment by jackpoz** [comment]
10s might be quite low, like in cases of Windows Updates randomly starting and using the whole bandwidth. Maybe 1 minute is high enough anyway ?

**Comment by Treeston** [comment]
You'd need to not send _any_ packets for 10s to kill the connection - that's not really something that's gonna happen just because some other program is using bandwidth.

**Comment by jackpoz** [comment]
It just has to happen once to trigger the disconnect, where now any client can have any kind of network lag spike without causing the disconnect (wifi signal degradation, mobile data connection with low performance, network saturation).
How often is the client sending a packet for sure anyway ? is there any operation on the computer that might trigger the disconnect (map loading on slow computer + other background unrelated processes, etc) ?

**Comment by Treeston** [comment]
Yes, but a client _shouldn't_ be able to stay connected for prolonged periods of time without data - the WoW client/server model simply isn't designed to handle it gracefully.

Keeping a player connected in these scenarios will make them visually teleport for others (impossible to fight against in PvP), as well as allowing them to clip through locked doors (as they aren't being sent in time), ignore NPC aggro (by moving from one side of the aggro range to the other before the next movement packet is sent) and many many more. It's simply not sustainable while maintaining a consistent gameplay environment (which is why the retail servers would act like this PR does).

**Comment by Shauren** [comment]
And 10s is too short (thats both CMSG_PING and CMSG_KEEPALIVE are on 30s timers)

**Comment by Treeston** [comment]
40s then?

**Comment by Shauren** [comment]
1 minute suggested earlier by jackpoz seems like a good idea to me

File: src/server/game/Server/WorldSocket.cpp [hunk 1]
```diff
@@ -358,6 +358,8 @@ WorldSocket::ReadDataHandlerResult WorldSocket::ReadDataHandler()
         }
         case CMSG_KEEP_ALIVE:
             LogOpcodeText(opcode, sessionGuard);
+            if (_worldSession)
```
Comment (Shauren) [inline src/server/game/Server/WorldSocket.cpp]: This is very bad - you are accessing _worldSession while sessionGuard is not locked

**Review by Shauren — CHANGES_REQUESTED** [review]

File: src/server/game/Server/WorldSocket.cpp (hunk 1 above)
Comment (Treeston) [inline src/server/game/Server/WorldSocket.cpp]: The only modification being made is to a `std::atomic` though. Still an issue?

Comment (Shauren) [inline src/server/game/Server/WorldSocket.cpp]: its the session pointer in WorldSocket that needs to be protected, not any member of WorldSession

Comment (jackpoz) [inline src/server/game/Server/WorldSocket.cpp]: if a race condition can happen here on _worldSession then its value could become NULL after the NULL-check, causing a NULL dereference crash.
if that's the case then it's just a matter of calling sessionGuard.lock() and then sessionGuard.unlock() (or whatever is the syntax of this lock) in the CMSG_KEEP_ALIVE case ?

Comment (Treeston) [inline src/server/game/Server/WorldSocket.cpp]: Makes sense, thanks @Shauren - I'll adjust this to lock/unlock.

Comment (Shauren) [inline src/server/game/Server/WorldSocket.cpp]: @jackpoz the type of sessionGuard is unique_lock (constructed with deferred lock flag), there is no need to unlock it

**Comment by Treeston** [comment]
@Shauren any other feedback or can this be merged?

**Review by Shauren — APPROVED** [review]

**Pull request merged**